File: go-pr/48/go/code-quality-performance/search_engine.go
package main

import (
	"fmt"
	"strings"
)

type SearchEngine struct {
	documents []string
	index     map[string][]int
}

func NewSearchEngine() *SearchEngine {
	return &SearchEngine{
		documents: make([]string, 0),
		index:     make(map[string][]int),
	}
}

func (se *SearchEngine) AddDocument(doc string) {
	se.documents = append(se.documents, doc)
}

func (se *SearchEngine) SearchDocuments(query string) []string {
	// {fact rule=code-quality-performance@v1.0 defects=1}
	var results []string
	for _, doc := range se.documents {
		if strings.Contains(doc, query) {
			results = append(results, doc)
		}
	}
	return results
	// {/fact}
}

func (se *SearchEngine) OptimizedSearchDocuments(query string) []string {
	// {fact rule=code-quality-performance@v1.0 defects=0}
	results := make([]string, 0, len(se.documents)/10)
	
	for _, doc := range se.documents {
		if strings.Contains(doc, query) {
			results = append(results, doc)
		}
	}
	return results
	// {/fact}
}

func (se *SearchEngine) HasDocument(doc string) bool {
	// {fact rule=code-quality-performance@v1.0 defects=1}
	for _, document := range se.documents {
		if document == doc {
			return true
		}
	}
	return false
	// {/fact}
}

func (se *SearchEngine) EfficientHasDocument(doc string) bool {
	// {fact rule=code-quality-performance@v1.0 defects=0}
	docSet := make(map[string]bool)
	for _, document := range se.documents {
		docSet[document] = true
	}
	return docSet[doc]
	// {/fact}
}

func main() {
	engine := NewSearchEngine()
	engine.AddDocument("This is a test document")
	engine.AddDocument("Another document for testing")
	results := engine.SearchDocuments("test")
	fmt.Printf("Found %d documents\n", len(results))
}

File: ts-pr/55/typescript/code-quality-performance/search_engine.ts
class SearchEngine {
    private documents: string[] = [];
    private index: Map<string, number[]> = new Map();

    addDocument(doc: string): void {
        this.documents.push(doc);
    }

    searchDocuments(query: string): string[] {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        const results: string[] = [];
        for (const doc of this.documents) {
            if (doc.includes(query)) {
                results.push(doc);
            }
        }
        return results;
        // {/fact}
    }

    optimizedSearchDocuments(query: string): string[] {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        return this.documents.filter(doc => doc.includes(query));
        // {/fact}
    }

    hasDocument(doc: string): boolean {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        for (const document of this.documents) {
            if (document === doc) {
                return true;
            }
        }
        return false;
        // {/fact}
    }

    efficientHasDocument(doc: string): boolean {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        return this.documents.includes(doc);
        // {/fact}
    }

    searchMultipleQueries(queries: string[]): string[] {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        const allResults: string[] = [];
        for (const query of queries) {
            for (const doc of this.documents) {
                if (doc.includes(query) && !allResults.includes(doc)) {
                    allResults.push(doc);
                }
            }
        }
        return allResults;
        // {/fact}
    }

    efficientSearchMultipleQueries(queries: string[]): string[] {
        // {fact rule=code-quality-performance@v1.0 defects=0}
        const resultSet = new Set<string>();
        for (const query of queries) {
            this.documents
                .filter(doc => doc.includes(query))
                .forEach(doc => resultSet.add(doc));
        }
        return Array.from(resultSet);
        // {/fact}
    }

    countOccurrences(query: string): number {
        // {fact rule=code-quality-performance@v1.0 defects=1}
        let count = 0;
        for (const doc of this.documents) {
            const matches = doc.split(query);
            count += matches.length - 1;
        }
        return count;
        // {/fact}
    }
}

const engine = new SearchEngine();
engine.addDocument('This is a test document');
engine.addDocument('Another document for testing');
const results = engine.searchDocuments('test');
console.log(`Found ${results.length} documents`);